### PR TITLE
[wip] bbolt backend

### PR DESCRIPTION
**Pull request by ecordell**
**Description of the change:**

- [x] bbolt data model
- [x] bbolt migrator
- [x] tests run against old and new backends
- [ ] finalize bbolt data model
- [ ] benchmark queries
- [ ] make queries efficient

**Motivation for the change:**

This will allow a pure-go build of registry tooling.

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/migrate/migrator.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+	"context"
+
+	"github.com/operator-framework/operator-registry/pkg/migrate/migrations"
+)
+
+type Migrator interface {
+	Migrate(ctx context.Context) error
+	Up(ctx context.Context, migrations migrations.Migrations) error
+	Down(ctx context.Context, migrations migrations.Migrations) error
+}
```
Comment (njhale) [inline pkg/migrate/migrator.go]: nit: missing newline.

File: pkg/sqlite/remove_test.go [hunk 2]
```diff
@@ -1,65 +1,65 @@
 package sqlite
 
-import (
-	"context"
-	"testing"
+// import (
```
Comment (njhale) [inline pkg/sqlite/remove_test.go]: Just a reminder that these tests are still commented out.

File: pkg/boltdb/migrate.go [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+func EnsureBolt(file string, backupFile string) error {
+	db, err := bolt.Open(file, 0600, nil)
+
+	// file is already a boltdb, no migration from sqlite needed
+	if err == nil {
+		return db.Close()
+	}
+
+	// move current db to a backup location
+	if err := os.Rename(file, backupFile); err != nil {
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/boltdb/migrate.go]: That's effectively what this does - moves the sqlite file to `.db.bak`, reads from the `.db.bak`, and writes to the `.db`

File: pkg/boltdb/migrate_test.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+func TestEnsureBolt(t *testing.T) {
+	prefix := fmt.Sprintf("test-%d", rand.Int())
+	dbName := prefix + ".db"
+	backupName := dbName + ".bak"
+	db, err := sql.Open("sqlite3", dbName)
+	require.NoError(t, err)
+	defer os.Remove(dbName)
+	defer os.Remove(backupName)
+
+	sqliteLoader, err := sqlite.NewSQLLiteLoader(db)
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/boltdb/migrate_test.go]: agreed

File: pkg/boltdb/query.go [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+		PackageName:  pkgName,
+		ChannelName:  channelName,
+		ProvidedApis: []*api.GroupVersionKind{},
+		RequiredApis: []*api.GroupVersionKind{},
+	}
+
+	// provided apis
+	for _, cap := range opBundle.Capabilities {
+		fmt.Printf("%#v", cap)
+		gvkp := strings.Split(cap.Value, "/")
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/boltdb/query.go]: if we keep this field as a string, yes

**Pull request closed without merging**